test/unit/puppetlabs/services/jruby/jruby_puppet_service_test.clj

```diff
@@ -289,3 +289,128 @@
+                            service
+                            "production")))
+         (is (= "1234test" (jruby-protocol/get-environment-class-info-tag
+                            service
+                            "test"))))
+       (let [production-first-update
+             (jruby-protocol/get-environment-class-info-tag-last-updated
+              service
+              "production")
+             test-first-update
+             (jruby-protocol/get-environment-class-info-tag-last-updated
+              service
+              "test")]
+         (testing "when environment info reset in the cache"
+           (is (not (nil? production-first-update)))
+           (is (not (nil? test-first-update)))
+           (jruby-protocol/set-environment-class-info-tag!
+            service
+            "production"
+            "5678prod"
+            production-first-update)
+           (is (= "5678prod" (jruby-protocol/get-environment-class-info-tag
+                              service
+                              "production")))
+           (is (= "1234test" (jruby-protocol/get-environment-class-info-tag
+                              service
+                              "test")))
+           (is (= test-first-update
+                  (jruby-protocol/get-environment-class-info-tag-last-updated
+                   service
+                   "test")))
+           (testing "and environment expired between get and corresponding set"
+             (let [production-second-update
+                   (jruby-protocol/get-environment-class-info-tag-last-updated
+                    service
+                    "production")
+                   _ (jruby-protocol/mark-environment-expired! service
+                                                               "production")
+                   production-third-update
+                   (jruby-protocol/get-environment-class-info-tag-last-updated
+                    service
+                    "production")]
+               (is (not= production-first-update production-second-update))
+               (jruby-protocol/set-environment-class-info-tag!
+                service
+                "production"
+                "89abprod"
+                production-second-update)
+               (is (= nil (jruby-protocol/get-environment-class-info-tag
+                           service
+                           "production")))
+               (is (= production-third-update
+                      (jruby-protocol/get-environment-class-info-tag-last-updated
+                       service
+                       "production"))))))
+         (testing "when an individual environment is marked expired"
+           (jruby-protocol/mark-environment-expired! service "production")
+           (is (nil? (jruby-protocol/get-environment-class-info-tag
+                      service
+                      "production")))
+           (is (not= production-first-update
```

Comment:
I find it confusing that this is `(not= production-first-update` rather than `(not = production-third-update` because that's the last update timestamp we have, although I guess that's because `production-third-update` isn't in scope here. It might be less confusing to get another `production-<x>-update` after line 379 here to compare against instead?